### ISSUE TITLE
Add option to ignore nested attachments

### ISSUE THREAD

**Issue by couteau** [php-mime-mail-parser/php-mime-mail-parser#270]
The getAttachments method of the Parser class currently returns all attachments within the MIME structure, even if they are actually attached to another (embedded email) attachment rather than directly to the top level MIME message. I sometimes need to get only the attachments that are the immediate children of the top level message (and then to separately parse any embedded/attached rfc822 messages to extract their attachments). I couldn't figure out a way to do this with the existing classes.

The following patch adds a parameter to the getAttachments method to ignore sub-attachments. It includes a default value that keeps the current behavior. 

```diff
--- a/php-mime-mail-parser/src/Parser.php	2020-01-01 14:29:36.000000000 -0500
+++ b/php-mime-mail-parser/src/Parser.php	2019-12-31 17:32:00.000000000 -0500
@@ -501,14 +501,14 @@
      *
      * @return Attachment[]
      */
-    public function getAttachments($include_inline = true)
+    public function getAttachments($include_inline = true, $include_subparts = true)
     {
         $attachments = [];
         $dispositions = $include_inline ? ['attachment', 'inline'] : ['attachment'];
         $non_attachment_types = ['text/plain', 'text/html'];
         $nonameIter = 0;
 
-        foreach ($this->parts as $part) {
+        foreach ($this->parts as $partId => $part) {
             $disposition = $this->getPart('content-disposition', $part);
             $filename = 'noname';
 
@@ -533,6 +533,9 @@
             }
 
             if (in_array($disposition, $dispositions) === true) {
+                if (!$include_subparts && $this->partIdIsChildOfAnAttachment($partId)) {
+                    continue;
+                }
                 if ($filename == 'noname') {
                     $nonameIter++;
                     $filename = 'noname'.$nonameIter;
```

**Comment by eXorus** [comment]
Hi,

Thanks for the suggestion but the existing function signature is not very good so I don't want to add another boolean to configure the function.

```php
$attachments = $parser->getAttachments();
// return an array of all attachments (include inline attachments)

$attachments = $parser->getAttachments(false);
// return an array of all attachments (exclude inline attachments)
```

The good way to add your great suggestion is to have maybe another function, or an array, or const....

**Comment by couteau** [comment]
I'm not sure I completely understand what you had in mind, but here is another approach - it uses consts to define flags for the different attachment types and the parameter to getAttachments is a combination of bits indicating which types should be returned. The default is all attachments, which is the equivalent of the current default. Most existing code should continue to work -- explicitly passed true or false parameters will still pull in nested attachments. The only issue would be if someone is using a truthy or falsy expression for the parameter that is not a true boolean value. 

```diff
diff --git a/src/Parser.php b/src/Parser.php
index 6502b57..ed63f91 100644
--- a/src/Parser.php
+++ b/src/Parser.php
@@ -19,6 +19,10 @@ class Parser
     const ATTACHMENT_DUPLICATE_SUFFIX = 'DuplicateSuffix';
     const ATTACHMENT_RANDOM_FILENAME  = 'RandomFilename';
 
+    const ATTACHMENT_TYPE_INLINE = 1;
+    const ATTACHMENT_TYPE_SUBPART = 2;
+    const ATTACHMENT_TYPE_ALL = 3;
+
     /**
      * PHP MimeParser Resource ID
      *
@@ -501,14 +505,17 @@ class Parser
      *
      * @return Attachment[]
      */
-    public function getAttachments($include_inline = true)
+    public function getAttachments($attachment_types = self::ATTACHMENT_TYPE_ALL)
     {
+        $include_inline = $attachment_types & self::ATTACHMENT_TYPE_INLINE;
+        $include_subparts = $attachment_types & self::ATTACHMENT_TYPE_SUBPART || is_bool($attachment_types);
+
         $attachments = [];
         $dispositions = $include_inline ? ['attachment', 'inline'] : ['attachment'];
         $non_attachment_types = ['text/plain', 'text/html'];
         $nonameIter = 0;
 
-        foreach ($this->parts as $part) {
+        foreach ($this->parts as $partId => $part) {
             $disposition = $this->getPart('content-disposition', $part);
             $filename = 'noname';
 
@@ -533,6 +540,9 @@ class Parser
             }
 
             if (in_array($disposition, $dispositions) === true) {
+                if (!$include_subparts && $this->partIdIsChildOfAnAttachment($partId)) {
+                    continue;
+                }
                 if ($filename == 'noname') {
                     $nonameIter++;
                     $filename = 'noname'.$nonameIter;
```

**Comment by eXorus** [comment]
Thanks it's far better, could you do a PR and add some tests?

**Comment by sanmai** [comment]
Is this the issue fixed by #287?

**Comment by eXorus** [comment]
Yes
